List of updated override files.

Anu1601CS opened the pull request:
Pull Request for Issue #29 
### Summary of Changes
The new interface of updated override files.

### Expected result
A view when list is found.

![screenshot from 2018-07-10 00-00-55](https://user-images.githubusercontent.com/30978328/42471940-323d28a4-83dd-11e8-8a7d-810f14466d04.png)

![screenshot from 2018-07-10 00-02-02](https://user-images.githubusercontent.com/30978328/42471932-2b5e1bc4-83dd-11e8-9672-b047a65ea788.png)

A view when nothing change.

![screenshot from 2018-07-10 01-06-34](https://user-images.githubusercontent.com/30978328/42472024-87fb9ffa-83dd-11e8-85bc-8e9ded370e62.png)

**Features added.**
1.  A view of all updated override which shows files in the list which do belong to the current template.
2. Name of the updated file its with location.
3. When you click on any name you will redirect to edit view of that fille.
4. Location, name of template, modified date and update action.
5. An option see  https://github.com/joomla-projects/gsoc18_override_management/pull/30#issuecomment-403375757

![screenshot from 2018-07-10 01-14-23](https://user-images.githubusercontent.com/30978328/42472360-a7113d04-83de-11e8-8f45-9d695a1279fd.png)

astridx commented:
The first thing I noticed. If I delete an override file in a template, this is still shown in the view.

Because we do not want to change the list later, that's OK. But it confused me. I searched the file and thought it is a mistake.

How complicated is it to show only the files that are still present as an override in the template?

astridx commented:
@Anu1601CS 
I have just made a Joomla update. That worked correctly. The list of files was always correct. After that I installed an extension. Then I saw in the view that in the last update no override files were changed. That confused me. 

Reason for this is: At the moment we overwrite last list entries with each update.

I think we should save these files for every update and every installation and sort them by date in descending order. 
For each update there should be an entry displaying 
the updated override 
- files  (as we do now), 
- the version number of the extension or Joomla and
- the name of the update (extension or Joomla) described.

@Anu1601CS What do you think about this?

Anu1601CS commented:
> @Anu1601CS
> I have just made a few Joomla updates. That worked correctly. The list of files was always correct. After that I installed an extension. Then I saw in the view that in the last update no override files were changed. That confused me.
> 
> Reason for this is: At the moment we overwrite last list entries with each update.
> 
> I think we should save these files for every update and every installation and sort them by date in descending order.
> For each update there should be an entry displaying
> the updated override
> 
> files (as we do now),
> the version number of the extension or Joomla and
> the name of the update (extension or Joomla) described.
> @Anu1601CS What do you think about this?

I am not sure about this. I think it may be over functionality to store file information about each update. Which make a big array and a very big list to show. 

We are currently showing the last update changed files. And this is fine.

laoneo commented:
I would only show files in the list which do belong to the current template.

Anu1601CS commented:
> I would only show files in the list which do belong to the current template.

Ok @laoneo 
But, if user have too many templates with too many overrides. Then, he need to go through all template and check updated overrides one by one.

But, I think this is also a fact that the user has only interest in his default or current template.

Anu1601CS commented:
@laoneo Can we add an option whether to show all updated files or only show files in the list which do belong to the current template. ?

astridx commented:
> Same here we can add an option like the number of previous update information the user wants to store and displayed them. Mean, if user save let say 5 then list will show and save previous 5 updates information ( updated files) and display them. ?

Yes, that's a very good idea. Maybe we should set the default value to 5.

astridx commented:
> @astridx 
 Now you can see only files which are present in template override.
And, added an option to redirect to file from list

Many Thanks. I think thats great.

Anu1601CS commented:
@astridx You also agree to show files in the list which do belong to the current template ?

dimnax commented:
Yes. With current template I mean the template the view is opened (not the default template)

astridx commented:
> @astridx You also agree to show files in the list which do belong to the current template ?
Yes

astridx commented:
> Same  here we can add an option like the number of previous update information the user wants to store and displayed them. Mean, if user save let say 5 then list will show and save previous 5 updates information ( updated files) and display them. ?

That's the only thing left, right?

Anu1601CS commented:
> Same here we can add an option like the number of previous update information the user wants to store and displayed them. Mean, if user save let say 5 then list will show and save previous 5 updates information ( updated files) and display them. ?
> 
> That's the only thing left, right?

Yes

Anu1601CS commented:
> Same here we can add an option like the number of previous update information the user wants to store and displayed them. Mean, if user save let say 5 then list will show and save previous 5 updates information ( updated files) and display them. ?

@laoneo is this feature also fine for you ?

laoneo commented on administrator/components/com_templates/Model/TemplateModel.php:
```diff
@@ -233,6 +232,27 @@ public function prepareCoreFiles($dir, $client, $element, $template)
+	 */
+	public function getPluginState()
+	{
+		$db = $this->getDbo();
```
Ca we not use here PluginHelper::getPlugin?

astridx replied:
I think you can use Joomla\CMS\Plugin\PluginHelper;

Please try in file /administrator/components/com_templates/View/Template/HtmlView.php in line 147 this:
$this->pluginState = PluginHelper::isEnabled('installer', 'override');

Do not forget 
use Joomla\CMS\Plugin\PluginHelper; 
in the head.

Anu1601CS replied:
@laoneo @astridx Thanks :)

laoneo commented on administrator/components/com_templates/tmpl/template/default.php:
```diff
@@ -303,6 +303,13 @@
 <?php echo $this->loadTemplate('description'); ?>
 <?php echo JHtml::_('bootstrap.endTab'); ?>
+
+<?php if ($this->pluginState['plg_installer_override']->enabled) : ?>
```
Does the PluginHelper class not provide a method for it?

Anu1601CS replied:
@laoneo As, I can see PluginHelper class not provide a method for it. But, if you know then how?

laoneo commented on installation/sql/mysql/joomla.sql:
```diff
@@ -662,7 +662,7 @@ INSERT INTO `#__extensions` (`extension_id`, `package_id`, `name`, `type`, `elem
 (488, 0, 'plg_sampledata_multilang', 'plugin', 'multilang', 'sampledata', 0, 1, 1, 0, '', '', 0, '0000-00-00 00:00:00', 0, 0, ''),
 (489, 0, 'plg_extension_namespacemap', 'plugin', 'namespacemap', 'extension', 0, 1, 1, 0, '', '{}', 0, '0000-00-00 00:00:00', 0, 0, ''),
-(490, 0, 'plg_installer_override', 'plugin', 'override', 'installer', 0, 1, 1, 1, '', '', 0, '0000-00-00 00:00:00', 4, 0, ''),
+(490, 0, 'plg_installer_override', 'plugin', 'override', 'installer', 0, 1, 1, 1, '', '{"numupdate":3,"overridefiles":"{}"}', 0, '0000-00-00 00:00:00', 4, 0, ''),
```
Why is here numupdate equals 3?

Anu1601CS replied:
Yes, @laoneo You can see previous  3 updates, changed files. And, You can change the plugin setting, that how many previous updates you want to save.  I set the default to 3 for this.

laoneo commented on plugins/installer/folderinstaller/folderinstaller.xml:
```diff
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="utf-8"?>
 <extension version="3.6" type="plugin" group="installer">
-	<name>PLG_INSTALLER_FOLDERINSTALLER</name>
```
Can you do that in a separate pr against the main repo?

laoneo commented on plugins/installer/urlinstaller/urlinstaller.xml:
```diff
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="utf-8"?>
 <extension version="3.6" type="plugin" group="installer">
-	<name>PLG_INSTALLER_URLINSTALLER</name>
+	<name>plg_installer_urlinstaller</name>
```
Same as above.

astridx commented:
Personal opinion of me. I would not show the first message. This unsettles new user. Experienced users know it anyway. And I would like to see a hint of the param in the plugin that set the number of saved array entries. Best the value itself.

![templates customise cassiopeia test administration 5](https://user-images.githubusercontent.com/9974686/42498751-57a3b9b4-842d-11e8-8a07-43fca650b060.png)

Anu1601CS commented:
> What I do not see here is how are the files cleared? 

@laoneo You want something which clear specific file from history ?

Anu1601CS commented:
@astridx https://github.com/joomla-projects/gsoc18_override_management/pull/30#issuecomment-403746142
 Done!

astridx commented:
> @astridx #30 (comment)
Done!

You are really fast. I can not keep up with pulling :)

But: I just remembered that the value could have been changed later.  And then the sentence may not be right. Maybe we should write:

_Override files are up to date with core files. Nothing is changed in last extension or Joomla update. Currently you save the last 3 changes._

I am sorry that I did not remember earlier.

Anu1601CS commented:
> @astridx #30 (comment)
> Done!
> 
> You are really fast. I can not keep up with pulling :)
> 
> But: I just remembered that the value could have been changed later. And then the sentence may not be right. Maybe we should write:
> 
> Override files are up to date with core files. Nothing is changed in last extension or Joomla update. Currently you save the last 3 changes.
> 
> I am sorry that I did not remember earlier.


Thanks, And just go and try the PR. You can change value and the sentence adopt that chnage :p

astridx commented:
Can we delete this too columns, because the value is all the time the same in the table?

![templates customise atum test administration 1](https://user-images.githubusercontent.com/9974686/42503347-b8f3b2d6-8438-11e8-83a3-f74231c4e731.png)

Anu1601CS commented:
https://github.com/joomla-projects/gsoc18_override_management/pull/30#issuecomment-403769184
Done! @astridx

astridx commented:
@Anu1601CS Thank you very much for your quick work. I tested everything. Most works well.

But now I have one problem. I hope I made no mistake because I am tired. But:

I changed the file /modules/mod_banners/tmpl/default.php 
I created an override in template cassiopeia for mod_banners and changed the file /templates/cassiopeia/html/mod_banners/default.php . 
Then I  made a joomla update via custom url https://www.jah-tz.de/downloads/core/gsoc18_override_management/gsoc18_update_list.xml
I expected that in cassiopeia the file is show in the list of the updated files. But no file is shown.

Did I make a mistake?

Anu1601CS commented:
> @Anu1601CS Thank you very much for your quick work. I tested everything. Most works well.
> 
> But now I have one problem. I hope I made no mistake because I am tired. But:
> 
> I changed the file /modules/mod_banners/tmpl/default.php
> I created an override in template cassiopeia for mod_banners and changed the file /templates/cassiopeia/html/mod_banners/default.php .
> Then I made a joomla update via custom url https://www.jah-tz.de/downloads/core/gsoc18_override_management/gsoc18_update_list.xml
> I expected that in cassiopeia the file is show in the list of the updated files. But no file is shown.
> 
> Did I make a mistake?
> 
> For a custom module while update of the module everything works fine. For me only the Joomla Update makes the problem

@astridx As, I can see `/modules/mod_banners/tmpl/default.php`   whether it belongs Site or Administrator .  https://www.jah-tz.de/downloads/core/gsoc18_override_management/gsoc18_update_list.xml   This update link does not have any update for `/modules/mod_banners/tmpl/default.php`   . So, as we expected file is not shown in the list of the updated files.

Anu1601CS commented:
Check out this package https://www.jah-tz.de/downloads/core/gsoc18_override_management/Joomla_4.0.1-dev-Development-Update_Package.zip

Anu1601CS commented:
This is my result. 

![screenshot from 2018-07-11 07-28-06](https://user-images.githubusercontent.com/30978328/42546687-133c0f6a-84dc-11e8-9793-303f5a8b4a5a.png)
![screenshot from 2018-07-11 07-27-49](https://user-images.githubusercontent.com/30978328/42546692-169b411c-84dc-11e8-92e9-c51784ebe0d6.png)

astridx commented:
> @astridx As, I can see /modules/mod_banners/tmpl/default.php whether it belongs Site or Administrator . https://www.jah-tz.de/downloads/core/gsoc18_override_management/gsoc18_update_list.xml This update link does not have any update for /modules/mod_banners/tmpl/default.php . So, as we expected file is not shown in the list of the updated files.

You are right. I did not check the update link. I expected the all files are changed ...

astridx commented:
> We don't need name, template, corefile to store it as a params right ?
So, we can remove these right ?

I also have not found any place where we need one of the three values stored. I think we do not need them in the value of the parameter.

Anu1601CS commented:
@astridx All done!

Anu1601CS commented:
`@Anu1601CS @zero-24 Do you know why are "Some checks were not successful"?` 
mean ? Which checks were not successful?

astridx commented:
I mean this:
![list of updated override files by anu1601cs pull request 30 joomla projects gsoc18_override_management](https://user-images.githubusercontent.com/9974686/42558988-1a362474-84f3-11e8-8fdc-07730d70c02d.png)

If you click on details you see, that there is a problem: "CSS or XPath element with '#menu-collapse-icon' was not found." I am sure it depends not on our changes, but I am not sure if I can merge.
![joomla projects gsoc18_override_management drone](https://user-images.githubusercontent.com/9974686/42558993-1fb38478-84f3-11e8-8c80-d9b3eb3d122d.png)

astridx commented:
@laoneo I would merge this branch, OK?

Anu1601CS commented:
Yes, system testing failing. I also don't know.

astridx commented:
I will merge this branch now. I tested it and I'm sure the conflicts have nothing to do with our changes. I think Allon meant with "What I do not see here is how the files are cleared?" where the list is deleted. But we only save a certain number in the array. That is OK. And if he meant something else, we can correct it later.

zero-24 commented:
usually rebooting the tests help. Just ping me the next time 😄 (and give me some more time to react :P)

astridx commented:
@zero-24 Thank you for the answer. So the conflict is probably not really important.
